src/memory.c

```diff
@@ -109,3 +116,196 @@ void memory_free_page(void *pageaddr) {
+        page_info = page_info->next;
+    }
+
+    //If no such page exists, grab another!
+    if (!page_info) {
+        console_printf("no suitable pages found, getting a new page\n");
+        kmalloc_get_page();
+        page_info = kmalloc_head;
+    }
+
+    //find some large enough gap
+    int slots_start_index = kmalloc_locate_sufficient_gap(page_info, slots_needed);
+    console_printf("slots_start_index = %d\n", slots_start_index);
+     
+    //fill in the front of the first slot with number of slots info
+    //phys addr is addr of page info plus size of page info plus slot offset * slot size
+    //needs to fit a memory address of 4 bytes
+    uint32_t first_slot_phys_addr = (uint32_t)page_info + sizeof(struct kmalloc_page_info) + (KMALLOC_SLOT_SIZE * slots_start_index);
+    console_printf("Found slot with first_slot_phys_addr = %d\n", first_slot_phys_addr);
+ 
+    //cast first_slot_phys_addr to uint16_t *, so that we can dereference it and fill it with slots needed, a 16 bit integer
+    *((uint16_t *)first_slot_phys_addr) = (uint16_t)slots_needed;
+
+    //mark these as malloc'd in kmalloc_page
+    int i;
+    for (i = 0; i < slots_needed; i++) {
+        page_info->free[slots_start_index + i] = 0;
+
+    }
+    
+    //find largest remaining gap
+    page_info->max_free_gap = kmalloc_get_largest_gap_size(page_info);
+    console_printf("\nLargest remaining gap is %d slots\n", page_info->max_free_gap);
+
+    //return a pointer to the useful memory area, which is sizeof(uint16_t) into the first slot.
+    return (void *)(first_slot_phys_addr + sizeof(uint16_t));
+}
+
+void kfree_mark_free(struct kmalloc_page_info *page_info, void *mem_loc) {
+    //the number of consecutive slots consumed is stored just ahead of the pointer given
+    uint16_t slots_consumed = *((uint16_t *)(mem_loc - sizeof(uint16_t)));
+    console_printf("Freeing %d blocks beginning at mem_loc %x\n", slots_consumed, mem_loc - sizeof(uint16_t));
+
+    //find offset for how far into the page memloc is in units of slots
+    int first_slot_index = ((uint32_t)mem_loc - (uint32_t)page_info - sizeof(uint16_t) - sizeof(struct kmalloc_page_info)) / KMALLOC_SLOT_SIZE;
+
+    //mark the slots which were consumed all free, starting with first_slot_index
+    int i;
+    for (i = 0; i < slots_consumed; i++) {
+        page_info->free[first_slot_index + i] = 1;
+    }
+}
+
+void kfree(void *to_free) {
+    if (!to_free) {
+        console_printf("kfree is being asked to free a NULL pointer\n");
+        return;
+    }
+
+    uint32_t to_free_as_int = (uint32_t)to_free;
+    struct kmalloc_page_info *page_info = kmalloc_head;
+    int is_freed = 0;
+    while (page_info && is_freed == 0) {
+        uint32_t page_start = (uint32_t)page_info;
+
+        //not <= and >= because pointers should never be on first or last byte of page
+        if ( page_start < to_free && page_start + PAGE_SIZE > to_free) {
```

Comment:
I think we need casts here, or warnings will be thrown about using pointers for arithmetic.